### ISSUE TITLE
Provide a way to escape special characters in field names

### ISSUE THREAD

**Issue by musicformellons** [mapbox/mapbox-gl-style-spec#362]
In a good working stylesheet I replaced {name} with existing fieldname {name:fr} (and also tried quoting it in various ways) but it is not recognized. See also: 
https://github.com/mapzen/vector-datasource/issues/105#issuecomment-142485513

**Comment by jfirebaugh** [comment]
We've reserved various punctuation characters, including the colon, for potential future use, e.g. [token defaults](https://github.com/mapbox/mapbox-gl-style-spec/issues/104). Mapbox vector sources use an underscore instead, e.g. `{name_fr}`.

**Comment by musicformellons** [comment]
ok, thx.

**Comment by nvkelso** [comment]
@jfirebaugh Is there any plan to allow {"name:fr"} to access the property named "name:fr", while still preserving the token defaults? 

Is there another existing technique to access a property besides the mustache syntax (like `feature["name:fr"]`?

**Comment by lucaswoj** [comment]
What's the ideal syntax for escaping special characters in field names? Some ideas:
- `{"name:fr"}`
- `{name\:fr}`
- `{[name:fr]}`

**Comment by musicformellons** [comment]
@nvkelso Could provide a suggestion that would work for you?

**Comment by nvkelso** [comment]
Either of these floats my boat (first one, quoted, is my preferred option):
- `{"name:fr"}` - this seems most straight forward and SQL-like
- `{[name:fr]}` - this is more like how Cascadenik used to do it, which is a nod to the past and optional.

This seems annoying:
- `{name\:fr}`

**Comment by musicformellons** [comment]
+1 for {"name:fr"}

**Comment by jfirebaugh** [comment]
Rather than extending our text-substitution mini-language with defaults and escaping, I'd rather abandon the mini-language entirely in favor of built-in JSON syntax. Something like:

```
"icon-image": {
  "type": "substitution",
  "value": [{"ref": "maki"}, "-12"]
}
```

Where elements in the "value" array are either strings or objects with a "ref" property whose value names a feature property key. This value would have no reserved characters or escaping.

Evaluation would proceed by replacing refs with stringified feature property values, leaving strings as literals, and then concatenating the resulting array of strings.

This extends to support (recursive) defaults nicely:

```
"icon-image": {
  "type": "substitution",
  "value": [{"ref": "maki", "default": default}, "-12"]
}
```

where `default` can be another string literal or ref object.

I don't recall the exact syntax we decided on for data-driven functions but essentially `"substitution"` should be just another function type.

**Comment by lucaswoj** [comment]
cc @1ec5

**Comment by 1ec5** [comment]
@jfirebaugh, your suggestion would be great, much better than a mini-language that will inevitably become as featureful and tricky to use as Bash variable operations. Would https://github.com/mapbox/mapbox-gl-style-spec/issues/104#issuecomment-159739278 require a different `type`, perhaps?

**Comment by lucaswoj** [comment]
@1ec5 We're open to syntax suggestions!

**Comment by nvkelso** [comment]
+1

**Comment by musicformellons** [comment]
+1

**Comment by planemad** [comment]
This is a blocker to render localized maps in studio using the `name:lang` tags exported from OSM. One needs to manually rename the tags in the data currently.

**Comment by musicformellons** [comment]
@lucaswoj Any news on this?

**Comment by sbma44** [comment]
this would also unlock some diagnostic mapping for the geocoder that relies on carmen properties which include colons

**Comment by lucaswoj** [comment]
@jfirebaugh @ansis Given that we don't want to go the route of a "mini-language" for token defaults, can we just lift the restrictions using colons in field names to unblock these use cases?

**Comment by jfirebaugh** [comment]
Good point -- yes, I think we can (with the exception of `{` and `}`, of course).

**Comment by lucaswoj** [comment]
Closing in favor of #419
